### PR TITLE
📦 Move commands to their own crates

### LINKED ISSUE #14: Extract the command handling functions as a trait
We need to have a trait for each command in the app. Since they are all isolated, common code can live elsewhere. This is to simplify testing and declutter `main.rs`

### PR DESCRIPTION
**Pull request by heytherewill**
Closes #14 

---

I ended up not using traits because the hassle was much bigger. Since the main feature of this was testability, now we do have that with properly extracted functions that can be unit tested because the components they need are injected in the  `execute` function, even if they don't implement a common trait.

This is best reviewed on a per-commit basis.

### CODE REVIEW

File: src/commands/cont.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            None => println!("{}", "No time entries in last 90 days".red()),
+            Some(time_entry) => {
+                let start = Utc::now();
+                let time_entry_to_create = TimeEntry {
+                    start,
+                    stop: None,
+                    duration: -start.timestamp(),
+                    created_with: Some(constants::CLIENT_NAME.to_string()),
+                    ..time_entry.clone()
+                };
```
Comment (shantanuraj) [inline src/commands/cont.rs:24]: Can we keep the helpers for starting and stop time-entries?

@tcrammond is working on the start command for #8 which could benefit from this common implementation and keeping things dry ☔

File: src/commands/auth.rs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    ) -> ResultWithDefaultError<()> {
+        let user = api_client.get_user().await?;
+        credentials_storage.persist(user.api_token)?;
+        println!(
+            "{} {}",
+            "Successfully authenticated for user with email:".green(),
+            user.email.green().bold(),
+        );
+
+        return Ok(());
```
Comment (shantanuraj) [inline src/commands/auth.rs:24]: image
clippy doesn't like this

Screenshot 2021-06-21 at 10 52 54


You can run

```shell
cargo clippy --fix -Z unstable-options
```

to autofix these, (needs nightly toolchain)

File: src/main.rs [hunk 3]
```diff
@@ -12,14 +14,17 @@ use arguments::Command::Running;
 use arguments::Command::Start;
 use arguments::Command::Stop;
 use arguments::CommandLineArguments;
-use chrono::Utc;
 use colored::Colorize;
-use credentials::Credentials;
-use models::{ResultWithDefaultError, TimeEntry};
+use commands::auth::AuthenticationCommand;
+use commands::cont::ContinueCommand;
```
Comment (shantanuraj) [inline src/main.rs:19]: `cont` sounds a little risque, could we use the full name `continue` like the `running` module?

Comment (heytherewill) [reply src/main.rs:19]: I tried, but continue is a reserved word so vscode autocompletion looked weird 😭 

But I guess we can live with that if the compiler doesn't complain

File: src/commands/cont.rs (hunk 1 above)
Comment (heytherewill) [reply src/commands/cont.rs:24]: Sure thing! We can move these to a `common` crate or something

File: src/main.rs (hunk 3 above)
Comment (shantanuraj) [reply src/main.rs:19]: ahh, alrighty 👍

File: src/commands/cont.rs (hunk 1 above)
Comment (shantanuraj) [reply src/commands/cont.rs:24]: That works 🚀

**Comment by heytherewill** [comment]
@shantanuraj I opted for keeping the methods as part of the TE impl, since this is easier to both discover and test. Let me know your thoughts

**Review by shantanuraj — APPROVED** [review]
tenor

**Pull request merged**